staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,233 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	golangproto "github.com/golang/protobuf/proto"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilresponsewriter "k8s.io/apiserver/pkg/util/responsewriter"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+// SchemaCacher fetches the openapi schema once and then caches it in memory
+type SchemaCacher struct {
```

Comment:
I'll try to address these three points separately, but also I am open to making changes to this code.

1: I agree that ideally the handler and path would be specified at creation time for the spec cacher. Unfortunately the way the apiserver works is that it installs all the api handlers first, then in PrepareRun (which runs after all installers finish) it builds and installs a handler for the openapi spec. This is necessary because the openapi spec's verbs are not known until all the installers have finished (although we don't care about the verbs for this, it does mean the openapi handler can't be installed until last).

2: The downloading here is done with the read lock taken, but I see what you mean that it might not need it the whole time. I want to think about that more.

3: The private methods which I think you are referring to (actually everything from line 138 down) is copied from the apply code in kubectl. I didn't change any of those methods at all because I was worried it might affect the behavior. Again, I am definitely open to making changes to it but here I just moved (well copied) that code to a new location.

If you meant the private method downloadSchema, its return values are the same as the values returned by the DownloadSchema method the aggregator uses when making an in process call to the openapi handler.

Reply:
Oh I think I see the benefit of setting the handler and path when the cacher is created,

Is it that we will never want to change those anyway, so there is no reason to make them mutable? (and making them immutable would simplify the locking as well)

Reply:
Okay I've recommitted with openAPIHandler and openAPIPath as an invariant and made a comment about them being immutable, and the read lock only being taken once in updateSchema to get the lastEtag

Reply:
I think it is definitely more understandable now :)